cmd/shfmt/main.go

```diff
@@ -240,10 +236,15 @@ func formatPath(path string, checkShebang bool) error {
 			}
 		}
 		if *write {
-			if err := f.Truncate(0); err != nil {
+			fi, err := f.Stat()
+			if err != nil {
+				return err
+			}
+			if err := f.Close(); err != nil {
 				return err
 			}
-			if _, err := f.Seek(0, io.SeekStart); err != nil {
+			f, err = os.OpenFile(path, os.O_WRONLY|os.O_TRUNC, fi.Mode())
```

Comment:
I'm fairly positive that you don't need to pass a real value for `perm` here. That is only used for `O_CREATE` - i.e., when opening a file for writing and the file doesn't exist.

For example, from `man 2 open` on linux:

> The  mode  argument  specifies  the file mode bits be applied when a new file is created.  This argument must be supplied when O_CREAT or O_TMPFILE is specified in flags; if neither O_CREAT nor O_TMPFILE is specified, then mode is ignored.

You could say "what if the file is deleted between the close and open calls", which is the kind of silliness you have to deal with when you want to open the same file twice. But I think that worrying about that technical possibility is not worth anyone's time. If the user is touching the files while the tool is running, lots of other things could go wrong anyway.

As an example - on the Go tree, doing `git grep 'OpenFile(' | grep WRONLY | grep -v CREATE`, I can see many `OpenFile(path, os.O_WRONLY, 0)` calls. I believe we can do the same.

Reply:
Good to know, will update my PR. I was actually checking if we need to pass to perms to the open call or not, but must have missed that section you quoted.

IMHO the "worst" that could happen due to opening the file twice is that the disk runs full between close and open and the content gets lost. But this wasn't better before.